File: src/sokoenginepy/tessellation/direction/direction.py
from enum import IntEnum
from typing import List


class Direction(IntEnum):
    """Directions of movement."""

    UP = 0
    NORTH_EAST = 1
    RIGHT = 2
    SOUTH_EAST = 3
    DOWN = 4
    SOUTH_WEST = 5
    LEFT = 6
    NORTH_WEST = 7

    def __repr__(self):
        return "Direction." + self.name

    @property
    def opposite(self) -> "Direction":
        return _OPPOSITE[self.value]


_OPPOSITE: List[Direction] = [
    Direction.DOWN,
    Direction.SOUTH_WEST,
    Direction.LEFT,
    Direction.NORTH_WEST,
    Direction.UP,
    Direction.NORTH_EAST,
    Direction.RIGHT,
    Direction.SOUTH_EAST,
]

File: src/libsokoengine/common/direction.cpp
/// @file
#include "direction.hpp"

#include <boost/algorithm/string.hpp>

using std::runtime_error;
using std::string;

namespace sokoengine {
namespace implementation {

string to_str(Direction d) {
  switch (d) {
    case Direction::UP:
      return "Direction.UP";
      break;
    case Direction::DOWN:
      return "Direction.DOWN";
      break;
    case Direction::LEFT:
      return "Direction.LEFT";
      break;
    case Direction::RIGHT:
      return "Direction.RIGHT";
      break;
    case Direction::NORTH_EAST:
      return "Direction.NORTH_EAST";
      break;
    case Direction::NORTH_WEST:
      return "Direction.NORTH_WEST";
      break;
    case Direction::SOUTH_EAST:
      return "Direction.SOUTH_EAST";
      break;
    case Direction::SOUTH_WEST:
      return "Direction.SOUTH_WEST";
      break;
  }
  throw runtime_error("Unhandled direction!");
}

} // namespace implementation
} // namespace sokoengine
